rpk: supports triggering on-demand balancer

daisukebe opened the pull request:
Although the continuous balancer has been becoming smart, there's still needs where we need to invoke the on-demand balancer to balance partitions across brokers. This commit adds a new `rpk cluster partitions balance` command that wraps the `POST /v1/partitions/rebalance` endpoint.

Internal discussion: https://redpandadata.slack.com/archives/C04LYNJQUGN/p1722580784813209

## Backports Required

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v24.2.x
- [x] v24.1.x
- [ ] v23.3.x

## Release Notes

### Features

* rpk: supports triggering on-demand partition balancer

daisukebe commented:
I should have updated rpadmin first. Tagged v0.1.4 https://github.com/redpanda-data/common-go/tree/rpadmin/v0.1.4/rpadmin, but it's not in the `main` branch.

Deflaimun commented on src/go/rpk/pkg/cli/cluster/partitions/balancer_run.go:
```diff
@@ -0,0 +1,53 @@
+		Short: "Triggers on-demand partition balancer",
+		Long: `Triggers on-demand partition balancer.
+
+This command allows you to trigger on-demand partition balancer.
```
We probably need a longer _long_ description here. We could possibly explain why the command is needed, requirements (if any), and what's outcome after running this command. 

I would also add this part here:
```
To check the command progress run 'rpk cluster partitions balancer-status'
```

daisukebe replied:
Thank you. I will add more details in the long description.

My understanding is this command works independently from continuous balancer. 
@ztlpn, is it safe to run the on-demand balancer any time, i.e. when the continuous balancer is running? If not, probably it's better to hide the command by default.

ztlpn replied:
Yes, it is safe.

r-vasquez commented:
@daisukebe Can you please delete and recreate the tag from main once https://github.com/redpanda-data/common-go/pull/12 is merged?

daisukebe commented:
Force pushed: https://github.com/redpanda-data/redpanda/commit/72ee8ca97e8dd3b07c715ee5b44f8e778cc67802

* Updated to https://github.com/redpanda-data/common-go/tree/rpadmin/v0.1.4
* Added some detailed information in the long help text

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/52920#019151c9-88c8-4d35-923b-dc4aed2aa181
<!--9aeb5b28-e7fc-4af8-8efd-0ed382c0ce85-->

Feediver1 commented on src/go/rpk/pkg/cli/cluster/partitions/balancer_run.go:
```diff
@@ -0,0 +1,73 @@
+		Short: "Triggers on-demand partition balancer",
+		Long: `Triggers on-demand partition balancer.
+
+This command allows you to trigger on-demand partition balancer.
```
```suggestion
This command allows you to trigger on-demand partition balancing.
```

Feediver1 commented on src/go/rpk/pkg/cli/cluster/partitions/balancer_run.go:
```diff
@@ -0,0 +1,73 @@
+
+This command allows you to trigger on-demand partition balancer.
+
+With Redpanda Community Edition, partition count on each broker
+easily becomes uneven, which leads to data skew. In order to
+distribute partitions across brokers, user can run this command
+to trigger on-demand partition balancer.
+
+With Redpanda Enterprise Edition, even though Continuous Data Balancing
+monitors broker and rack availability, as well as disk usage, to avoid
+topic hotspots, there're still some edge cases where it can't handle well,
+i.e. a node becoming unavailable for a prolonged time and joining the
+cluster back thereafter. In such cases, user needs to run this command
+to trigger partition balancer manually.
+
+After you run this command, monitor the balancer progress using:
+
+    rpk cluster partitions balancer-status
+
+To see more detailed movement status, monitor the progress using:
+
+    rpk cluster partitions move-status
+`,
+
```
```suggestion
With Redpanda Community Edition, the partition count on each broker
can easily become uneven, which leads to data skewing. To
distribute partitions across brokers, you can run this command
to trigger on-demand partition balancing.

With Redpanda Enterprise Edition, Continuous Data Balancing
monitors broker and rack availability, as well as disk usage, to avoid
topic hotspots. However, there are edge cases where users should manually trigger partition balancing (such as a node becoming unavailable for a prolonged time and rejoining the
cluster thereafter). In such cases, you should run this command
to trigger partition balancing manually.

After you run this command, monitor the balancer progress using:

    rpk cluster partitions balancer-status

To see more detailed movement status, monitor the progress using:

    rpk cluster partitions move-status
`,

```

r-vasquez reviewed (dismissed):
Code LGTM, thanks!

twmb commented on src/go/rpk/pkg/cli/cluster/partitions/balancer_run.go:
```diff
@@ -0,0 +1,73 @@
+	"github.com/spf13/cobra"
+)
+
+func newTriggerBalnacerCommand(fs afero.Fs, p *config.Params) *cobra.Command {
```
nit: Balnacer => Balancer

twmb commented on src/go/rpk/pkg/cli/cluster/partitions/balancer_run.go:
```diff
@@ -0,0 +1,73 @@
+			out.MaybeDie(err, "unable to initialize admin client: %v", err)
+
+			err = cl.TriggerBalancer(cmd.Context())
+			out.MaybeDie(err, "failed to invoke ondemand partition balancer: %v", err)
```
Formatter is unused, I'd expect it to be ~here if it's in use.
cc @r-vasquez

r-vasquez replied:
Good catch, missed that during the review. I think we can remove the format flag from here, wdyt?

This is a command that is not expected to be automated and instead be run by an admin.

daisukebe replied:
Agreed. I will remove it. Thank you!

daisukebe commented:
Force pushed: https://github.com/redpanda-data/redpanda/commit/8164de9b3832f2da141b83e564a402ed1ff3f0e9

* Applied Joyce's suggestion on the help text
* Fixed the typo
* Removed the formatter

r-vasquez commented on src/go/rpk/pkg/cli/cluster/partitions/balancer_run.go:
```diff
@@ -0,0 +1,68 @@
+			fmt.Println("Successfully triggered on demand partition balancer.\n\nPlease find the progress with 'rpk cluster partitions balancer-status'.")
+		},
+	}
+	p.InstallFormatFlag(cmd)
```
This is not needed as you removed the format above.

r-vasquez replied:
@daisukebe just checking… I see that you resolved the comment but didn’t push the change

daisukebe replied:
My bad. I don't understand why I didn't change it. Thank you for catching i!

Deflaimun reviewed (dismissed):
lgtm from docs

daisukebe commented:
Force pushed again (sorry) to align with wording; `balancer` -> `balancing`.

r-vasquez reviewed (approved):
LGTM, thanks!

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
/backport v24.1.x

vbotbuildovich commented:
Failed to create a backport PR to v24.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-22855-v24.2.x-250 remotes/upstream/v24.2.x
git cherry-pick -x 62031165e3c6867939e182a4066bf6831088248a
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/10500471545).

vbotbuildovich commented:
Failed to create a backport PR to v24.1.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-22855-v24.1.x-736 remotes/upstream/v24.1.x
git cherry-pick -x 62031165e3c6867939e182a4066bf6831088248a
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/10500471478).
